include/ocpp/v201/smart_charging.hpp

```diff
@@ -0,0 +1,52 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2020 - 2023 Pionix GmbH and Contributors to EVerest
+
+#ifndef OCPP_V201_SMART_CHARGING_HPP
+#define OCPP_V201_SMART_CHARGING_HPP
+
+#include "ocpp/v201/enums.hpp"
+#include <limits>
+
+#include <ocpp/v201/database_handler.hpp>
+#include <ocpp/v201/evse.hpp>
+#include <ocpp/v201/ocpp_types.hpp>
+#include <ocpp/v201/transaction.hpp>
+
+namespace ocpp {
+namespace v201 {
+
+enum class ProfileValidationResultEnum {
+    Valid,
+    TxProfileMissingTransactionId,
+    TxProfileEvseIdNotGreaterThanZero,
+    TxProfileTransactionNotOnEvse,
+    TxProfileEvseHasNoActiveTransaction,
+    TxProfileConflictingStackLevel
+};
+
+/// \brief This class handles and maintains incoming ChargingProfiles and contains the logic
+/// to calculate the composite schedules
+class SmartChargingHandler {
+private:
+    std::shared_ptr<ocpp::v201::DatabaseHandler> database_handler;
+    // cppcheck-suppress unusedStructMember
+    std::vector<ChargingProfile> charging_profiles;
+
+public:
+    explicit SmartChargingHandler(std::shared_ptr<DatabaseHandler> database_handler);
+
+    ///
+    /// \brief validates the given \p profile according to the specification
+    ///
+    ProfileValidationResultEnum validate_tx_profile(ChargingProfile& profile, Evse& evse) const;
```

Comment:
I think these can be const references which probably indicates better what is going on. Same for `add_profile`

Reply:
We made the profile const, but the evse can't be const as the methods we use aren't const.

Reply:
That does make sense, but then again, you aren't actually changing anything so it might make sense to add the const functions you need? Not sure if that should be in scope of this PR though..